File: check-upstream-drift.py
#!/usr/bin/env python3
"""Report where the stored reference evidence no longer matches its upstream.

Every snapshot in this repository is dated and hashed: README blobs carry
`readme_blob_sha`, catalog entries carry `source_url` and `source_image_url`,
reference records carry a `source_url` per motion asset. Nothing re-checked any of
it, so a source that moved, went private, or changed content stayed invisible.

This utility performs three read-only checks and writes one report:

1. **README drift** — the current blob SHA of each snapshotted README, through the
   GitHub API (`gh`), against the SHA recorded at capture time.
2. **Source reachability** — an HTTP HEAD (falling back to a ranged GET) for every
   catalog `source_url`, `source_image_url`, and motion `source_url`.
3. **Local integrity** — every recorded local media path resolves and matches its
   recorded SHA-256.

    ./check-upstream-drift.py                     # everything, report to stdout
    ./check-upstream-drift.py --skip-network      # local integrity only
    ./check-upstream-drift.py --write-report      # also writes upstream-drift.json
    ./check-upstream-drift.py --strict            # exit 1 when drift is found
"""

from __future__ import annotations

import argparse
import hashlib
import json
import subprocess
import sys
import urllib.error
import urllib.request
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parent
REPORT = ROOT / "upstream-drift.json"
USER_AGENT = "WisentProductGuidelines/1.0 (+https://wisent.ai)"
TIMEOUT = 20
SCHEMA = "wisent.upstream-drift-report.v1"


@dataclass
class Drift:
    readme_changed: list[dict[str, Any]] = field(default_factory=list)
    readme_unreachable: list[dict[str, Any]] = field(default_factory=list)
    readme_unchanged: int = 0
    sources_gone: list[dict[str, Any]] = field(default_factory=list)
    sources_guarded: list[dict[str, Any]] = field(default_factory=list)
    sources_unresolved: list[dict[str, Any]] = field(default_factory=list)
    sources_ok: int = 0
    sources_skipped: int = 0
    media_missing: list[str] = field(default_factory=list)
    media_hash_mismatch: list[str] = field(default_factory=list)
    media_ok: int = 0

    def any_drift(self) -> bool:
        """Guarded sources are not drift: an authenticated product answering 401 is
        behaving as recorded. A gone or unresolvable source is drift."""
        return bool(
            self.readme_changed
            or self.readme_unreachable
            or self.sources_gone
            or self.sources_unresolved
            or self.media_missing
            or self.media_hash_mismatch
        )


def gh_json(path: str) -> dict[str, Any] | None:
    proc = subprocess.run(
        ["gh", "api", path, "--cache", "0"],
        capture_output=True,
        text=True,
    )
    if proc.returncode != 0:
        return None
    try:
        return json.loads(proc.stdout)
    except json.JSONDecodeError:
        return None


def check_readmes(drift: Drift) -> None:
    sources = json.loads((ROOT / "readme-examples" / "sources.json").read_text())
    entries = sources.get("repositories") or sources.get("examples") or sources.get("sources") or []

    def one(entry: dict[str, Any]) -> None:
        repo = entry.get("repository")
        readme_path = entry.get("readme_path")
        recorded = entry.get("readme_blob_sha")
        if not (repo and readme_path):
            return
        data = gh_json(f"repos/{repo}/contents/{readme_path}")
        if data is None or "sha" not in data:
            drift.readme_unreachable.append({"repository": repo, "readme_path": readme_path})
            return
        if data["sha"] != recorded:
            drift.readme_changed.append(
                {
                    "repository": repo,
                    "readme_path": readme_path,
                    "recorded_sha": recorded,
                    "current_sha": data["sha"],
                    "snapshot": entry.get("filename"),
                }
            )
        else:
            drift.readme_unchanged += 1

    with ThreadPoolExecutor(max_workers=8) as pool:
        list(pool.map(one, entries))


# A dead reference and a guarded one are different findings. An authenticated web
# app answering 401 at its dashboard URL is behaving exactly as recorded; a 404 means
# the page we cited is gone and the provenance is now wrong.
GONE_CODES = {400, 404, 410}
GUARDED_CODES = {401, 403, 405, 429, 451, 501, 503}


def url_state(url: str) -> tuple[str, int | str]:
    """Return (state, detail) where state is reachable | gone | guarded | unresolved."""
    request = urllib.request.Request(url, method="HEAD", headers={"User-Agent": USER_AGENT})
    try:
        with urllib.request.urlopen(request, timeout=TIMEOUT) as response:
            return "reachable", response.status
    except urllib.error.HTTPError as exc:
        if exc.code in GUARDED_CODES:  # HEAD refused or rate limited; try one byte
            try:
                ranged = urllib.request.Request(
                    url, headers={"User-Agent": USER_AGENT, "Range": "bytes=0-0"}
                )
                with urllib.request.urlopen(ranged, timeout=TIMEOUT) as response:
                    return "reachable", response.status
            except urllib.error.HTTPError as inner:
                if inner.code in GONE_CODES:
                    return "gone", inner.code
                return "guarded", inner.code
            except Exception as inner:  # noqa: BLE001 - reported, not raised
                return "guarded", f"{exc.code} then {type(inner).__name__}"
        if exc.code in GONE_CODES:
            return "gone", exc.code
        return "guarded", exc.code
    except urllib.error.URLError as exc:
        return "unresolved", type(exc.reason).__name__ if exc.reason else "URLError"
    except Exception as exc:  # noqa: BLE001 - network failure is the finding
        return "unresolved", type(exc).__name__


def collect_urls() -> list[tuple[str, str, str | None]]:
    """(url, where, expected_state) for every recorded upstream reference."""
    pairs: list[tuple[str, str, str | None]] = []
    for sources in sorted(ROOT.glob("*-examples/sources.json")):
        data = json.loads(sources.read_text())
        for example in data.get("examples", []) or []:
            name = example.get("name", "?")
            if example.get("source_url"):
                pairs.append((
                    example["source_url"],
                    f"{sources.parent.name}/{name}/source_url",
                    example.get("source_url_state"),
                ))
            visual = example.get("visual") or {}
            for key in ("source_page_url", "source_image_url"):
                if visual.get(key):
                    pairs.append((
                        visual[key],
                        f"{sources.parent.name}/{name}/{key}",
                        visual.get(f"{key}_state"),
                    ))
    for record in sorted(ROOT.glob("*-examples/references/*/reference.json")):
        data = json.loads(record.read_text())
        for entry in data.get("motion", []) or []:
            if entry.get("source_url"):
                pairs.append((
                    entry["source_url"],
                    f"{record.parent.parent.parent.name}/{record.parent.name}/motion",
                    entry.get("source_url_state"),
                ))
    seen: set[str] = set()
    unique: list[tuple[str, str, str | None]] = []
    for url, where, expected in pairs:
        if url in seen:
            continue
        seen.add(url)
        unique.append((url, where, expected))
    return unique


def check_sources(drift: Drift) -> None:
    buckets = {
        "gone": drift.sources_gone,
        "guarded": drift.sources_guarded,
        "unresolved": drift.sources_unresolved,
    }

    def one(pair: tuple[str, str, str | None]) -> None:
        url, where, expected = pair
        if not url.startswith(("http://", "https://")):
            drift.sources_skipped += 1
            return
        state, detail = url_state(url)
        if state == "reachable":
            drift.sources_ok += 1
            return
        # A private repository or authenticated application may deliberately
        # answer anonymous HTTP with 400/404. Its recorded classification wins
        # over the transport code; otherwise the checker reports a live guarded
        # surface as deleted on every run.
        if expected == "guarded" and state == "gone":
            state = "guarded"
        buckets[state].append({"url": url, "where": where, "result": detail, "expected": expected})

    with ThreadPoolExecutor(max_workers=12) as pool:
        list(pool.map(one, collect_urls()))


def check_local_media(drift: Drift) -> None:
    for record in sorted(ROOT.glob("*-examples/references/*/reference.json")):
        data = json.loads(record.read_text())
        base = record.parent
        for key in ("motion", "states"):
            for entry in data.get(key, []) or []:
                local = base / str(entry.get("local_path", ""))
                rel = str(local.relative_to(ROOT))
                if not local.exists():
                    drift.media_missing.append(rel)
                    continue
                recorded = entry.get("sha256")
                if not recorded:
                    continue
                digest = hashlib.sha256()
                with local.open("rb") as fh:
                    for chunk in iter(lambda: fh.read(1 << 20), b""):
                        digest.update(chunk)
                if digest.hexdigest() != recorded:
                    drift.media_hash_mismatch.append(rel)
                else:
                    drift.media_ok += 1


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--skip-network", action="store_true")
    parser.add_argument("--skip-readme", action="store_true")
    parser.add_argument("--write-report", action="store_true")
    parser.add_argument("--strict", action="store_true")
    args = parser.parse_args()

    drift = Drift()
    check_local_media(drift)
    if not args.skip_network:
        if not args.skip_readme:
            check_readmes(drift)
        check_sources(drift)

    print(f"local media verified: {drift.media_ok}")
    print(f"local media missing: {len(drift.media_missing)}")
    print(f"local media hash mismatch: {len(drift.media_hash_mismatch)}")
    if not args.skip_network:
        print(f"README snapshots unchanged: {drift.readme_unchanged}")
        print(f"README snapshots changed upstream: {len(drift.readme_changed)}")
        print(f"README snapshots unreachable: {len(drift.readme_unreachable)}")
        print(f"upstream URLs reachable: {drift.sources_ok}")
        print(f"upstream URLs gone: {len(drift.sources_gone)}")
        print(f"upstream URLs guarded (auth, rate limit, bot wall): {len(drift.sources_guarded)}")
        print(f"upstream URLs unresolved (network): {len(drift.sources_unresolved)}")
    for item in drift.media_missing[:20]:
        print(f"  missing media: {item}")
    for item in drift.media_hash_mismatch[:20]:
        print(f"  hash mismatch: {item}")
    for item in drift.readme_changed[:20]:
        print(f"  README changed: {item['repository']} ({item['snapshot']})")
    for item in drift.sources_gone:
        print(f"  upstream gone: {item['where']} -> {item['result']} {item['url']}")
    for item in drift.sources_unresolved:
        print(f"  upstream unresolved: {item['where']} -> {item['result']} {item['url']}")

    if args.write_report:
        REPORT.write_text(
            json.dumps(
                {
                    "schema": SCHEMA,
                    "checked_at": subprocess.run(
                        ["date", "-u", "+%Y-%m-%dT%H:%M:%SZ"], capture_output=True, text=True
                    ).stdout.strip(),
                    "network_checked": not args.skip_network,
                    "local_media_verified": drift.media_ok,
                    "local_media_missing": drift.media_missing,
                    "local_media_hash_mismatch": drift.media_hash_mismatch,
                    "readme_unchanged": drift.readme_unchanged,
                    "readme_changed": drift.readme_changed,
                    "readme_unreachable": drift.readme_unreachable,
                    "upstream_urls_reachable": drift.sources_ok,
                    "upstream_urls_gone": drift.sources_gone,
                    "upstream_urls_guarded": drift.sources_guarded,
                    "upstream_urls_unresolved": drift.sources_unresolved,
                    "upstream_urls_skipped": drift.sources_skipped,
                },
                indent=2,
            )
            + "\n"
        )
        print(f"\nreport written to {REPORT.name}")

    return 1 if args.strict and drift.any_drift() else 0


if __name__ == "__main__":
    sys.exit(main())

File: src/commands/check_upstream_drift.rs
//! `spis check-upstream-drift` — report where stored reference evidence no
//! longer matches its upstream. Read-only checks:
//!
//! 1. README drift — current blob SHA of each snapshotted README via the
//!    GitHub API (`gh` CLI, preserved external call) against the recorded SHA.
//! 2. Source reachability — HTTP HEAD falling back to a ranged GET for every
//!    catalog `source_url`, `source_image_url`, and motion `source_url`.
//! 3. Local integrity — every recorded local media path resolves and matches
//!    its recorded SHA-256.
//!
//! Ported 1:1 from the former check-upstream-drift.py.

use crate as lib;
use anyhow::{Context, Result};
use parking_lot::Mutex;
use serde_json::{json, Value};
use std::path::PathBuf;
use std::process::Command;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::time::Duration;

const REPORT: &str = "upstream-drift.json";
const USER_AGENT: &str = crate::USER_AGENT;
const SCHEMA: &str = "wisent.upstream-drift-report.v1";
const TIMEOUT_SECS: u64 = 20;

const GONE_CODES: &[u16] = &[400, 404, 410];
const GUARDED_CODES: &[u16] = &[401, 403, 405, 429, 451, 501, 503];

#[derive(Default)]
struct Drift {
    readme_changed: Vec<Value>,
    readme_unreachable: Vec<Value>,
    readme_unchanged: usize,
    sources_gone: Vec<Value>,
    sources_guarded: Vec<Value>,
    sources_unresolved: Vec<Value>,
    sources_ok: usize,
    sources_skipped: usize,
    media_missing: Vec<String>,
    media_hash_mismatch: Vec<String>,
    media_ok: usize,
}

impl Drift {
    /// Guarded sources are not drift: an authenticated product answering 401
    /// is behaving as recorded. A gone or unresolvable source is drift.
    fn any_drift(&self) -> bool {
        !self.readme_changed.is_empty()
            || !self.readme_unreachable.is_empty()
            || !self.sources_gone.is_empty()
            || !self.sources_unresolved.is_empty()
            || !self.media_missing.is_empty()
            || !self.media_hash_mismatch.is_empty()
    }
}

fn agent() -> ureq::Agent {
    ureq::AgentBuilder::new()
        .timeout(Duration::from_secs(TIMEOUT_SECS))
        .user_agent(USER_AGENT)
        .build()
}

/// Run `f` over `items` on a bounded worker pool (mirrors ThreadPoolExecutor).
fn pool<T: Sync, F: Fn(&T) + Sync>(items: &[T], workers: usize, f: F) {
    if items.is_empty() {
        return;
    }
    let next = AtomicUsize::new(0);
    std::thread::scope(|s| {
        for _ in 0..workers.min(items.len()) {
            s.spawn(|| loop {
                let i = next.fetch_add(1, Ordering::SeqCst);
                if i >= items.len() {
                    break;
                }
                f(&items[i]);
            });
        }
    });
}

fn gh_json(path: &str) -> Option<Value> {
    // Preserved external call: `gh api <path> --cache 0`.
    let out = Command::new("gh")
        .args(["api", path, "--cache", "0"])
        .output()
        .ok()?;
    if !out.status.success() {
        return None;
    }
    serde_json::from_slice(&out.stdout).ok()
}

fn check_readmes(drift: &Mutex<Drift>) -> Result<()> {
    let sources: Value = lib::read_json("readme-examples/sources.json")?;
    let entries = sources
        .get("repositories")
        .or_else(|| sources.get("examples"))
        .or_else(|| sources.get("sources"))
        .and_then(|v| v.as_array())
        .cloned()
        .unwrap_or_default();

    pool(&entries, 8, |entry| {
        let repo = entry.get("repository").and_then(|v| v.as_str());
        let readme_path = entry.get("readme_path").and_then(|v| v.as_str());
        let recorded = entry.get("readme_blob_sha");
        let (repo, readme_path) = match (repo, readme_path) {
            (Some(r), Some(p)) => (r, p),
            _ => return,
        };
        let mut guard = drift.lock();
        match gh_json(&format!("repos/{repo}/contents/{readme_path}")) {
            None => guard.readme_unreachable.push(json!({
                "repository": repo,
                "readme_path": readme_path,
            })),
            Some(data) => match data.get("sha").and_then(|v| v.as_str()) {
                None => guard.readme_unreachable.push(json!({
                    "repository": repo,
                    "readme_path": readme_path,
                })),
                Some(current) => {
                    if Some(current) != recorded.and_then(|v| v.as_str()) {
                        guard.readme_changed.push(json!({
                            "repository": repo,
                            "readme_path": readme_path,
                            "recorded_sha": recorded.cloned().unwrap_or(Value::Null),
                            "current_sha": current,
                            "snapshot": entry.get("filename").cloned().unwrap_or(Value::Null),
                        }));
                    } else {
                        guard.readme_unchanged += 1;
                    }
                }
            },
        }
    });
    Ok(())
}

/// Return (state, detail) where state is reachable | gone | guarded | unresolved.
fn url_state(client: &ureq::Agent, url: &str) -> (&'static str, Value) {
    match client.head(url).call() {
        Ok(resp) => ("reachable", json!(resp.status())),
        Err(ureq::Error::Status(code, resp)) => {
            if GUARDED_CODES.contains(&resp.status()) {
                // HEAD refused or rate limited; try one byte.
                match client.get(url).set("Range", "bytes=0-0").call() {
                    Ok(r) => return ("reachable", json!(r.status())),
                    Err(ureq::Error::Status(inner, _)) => {
                        if GONE_CODES.contains(&inner) {
                            return ("gone", json!(inner));
                        }
                        return ("guarded", json!(inner));
                    }
                    Err(_) => {
                        // Python reported "{code} then {ExceptionName}".
                        return (
                            "guarded",
                            json!(format!("{} then TransportError", resp.status())),
                        );
                    }
                }
            }
            if GONE_CODES.contains(&code) {
                return ("gone", json!(code));
            }
            ("guarded", json!(code))
        }
        Err(e) => ("unresolved", json!(format!("{}", e_kind(&e)))),
    }
}

fn e_kind(e: &ureq::Error) -> &'static str {
    match e {
        ureq::Error::Status(_, _) => "Status",
        ureq::Error::Transport(t) => match t.kind() {
            ureq::ErrorKind::Dns => "Dns",
            _ => "URLError",
        },
    }
}

/// (url, where, expected_state) for every recorded upstream reference.
fn collect_urls() -> Vec<(String, String, Option<String>)> {
    let mut pairs: Vec<(String, String, Option<String>)> = Vec::new();
    for sources in sorted_glob_examples_files("sources.json") {
        let Ok(data) = lib::read_json::<Value>(sources.to_str().unwrap()) else {
            continue;
        };
        let catalog = sources
            .parent()
            .and_then(|p| p.file_name())
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_default();
        for example in data
            .get("examples")
            .and_then(|v| v.as_array())
            .map(|a| a.as_slice())
            .unwrap_or(&[])
        {
            let name = example
                .get("name")
                .and_then(|v| v.as_str())
                .unwrap_or("?")
                .to_string();
            if let Some(url) = example.get("source_url").and_then(|v| v.as_str()) {
                pairs.push((
                    url.to_string(),
                    format!("{catalog}/{name}/source_url"),
                    example
                        .get("source_url_state")
                        .and_then(|v| v.as_str())
                        .map(str::to_string),
                ));
            }
            if let Some(visual) = example.get("visual") {
                for key in ["source_page_url", "source_image_url"] {
                    if let Some(url) = visual.get(key).and_then(|v| v.as_str()) {
                        pairs.push((
                            url.to_string(),
                            format!("{catalog}/{name}/{key}"),
                            visual
                                .get(&format!("{key}_state"))
                                .and_then(|v| v.as_str())
                                .map(str::to_string),
                        ));
                    }
                }
            }
        }
    }
    for record in reference_records() {
        let Ok(data) = lib::read_json::<Value>(record.to_str().unwrap()) else {
            continue;
        };
        let catalog = record
            .parent()
            .and_then(|p| p.parent())
            .and_then(|p| p.parent())
            .and_then(|p| p.file_name())
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_default();
        let dir_name = record
            .parent()
            .and_then(|p| p.file_name())
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_default();
        for entry in data
            .get("motion")
            .and_then(|v| v.as_array())
            .map(|a| a.as_slice())
            .unwrap_or(&[])
        {
            if let Some(url) = entry.get("source_url").and_then(|v| v.as_str()) {
                pairs.push((
                    url.to_string(),
                    format!("{catalog}/{dir_name}/motion"),
                    entry
                        .get("source_url_state")
                        .and_then(|v| v.as_str())
                        .map(str::to_string),
                ));
            }
        }
    }
    let mut seen = std::collections::HashSet::new();
    pairs.retain(|(url, _, _)| seen.insert(url.clone()));
    pairs
}

fn check_sources(drift: &Mutex<Drift>) {
    let pairs = collect_urls();
    pool(&pairs, 12, |pair| {
        let (url, where_, expected) = pair;
        if !(url.starts_with("http://") || url.starts_with("https://")) {
            drift.lock().sources_skipped += 1;
            return;
        }
        let client = agent();
        let (mut state, detail) = url_state(&client, url);
        if state == "reachable" {
            drift.lock().sources_ok += 1;
            return;
        }
        // A private repository or authenticated application may deliberately
        // answer anonymous HTTP with 400/404. Its recorded classification wins
        // over the transport code.
        if expected.as_deref() == Some("guarded") && state == "gone" {
            state = "guarded";
        }
        let item = json!({
            "url": url,
            "where": where_,
            "result": detail,
            "expected": expected.clone().map(Value::String).unwrap_or(Value::Null),
        });
        let mut guard = drift.lock();
        match state {
            "gone" => guard.sources_gone.push(item),
            "guarded" => guard.sources_guarded.push(item),
            _ => guard.sources_unresolved.push(item),
        }
    });
}

fn stream_sha256(path: &std::path::Path) -> Result<String> {
    use sha2::Digest;
    use std::io::Read;
    let mut file = std::fs::File::open(path)?;
    let mut digest = sha2::Sha256::new();
    let mut buf = vec![0u8; 1 << 20];
    loop {
        let n = file.read(&mut buf)?;
        if n == 0 {
            break;
        }
        digest.update(&buf[..n]);
    }
    Ok(digest
        .finalize()
        .iter()
        .map(|b| format!("{b:02x}"))
        .collect())
}

fn check_local_media(drift: &mut Drift) -> Result<()> {
    for record in reference_records() {
        let data: Value = lib::read_json(record.to_str().context("non-UTF8 path")?)?;
        let base = record.parent().context("record without parent")?;
        for key in ["motion", "states"] {
            for entry in data
                .get(key)
                .and_then(|v| v.as_array())
                .map(|a| a.as_slice())
                .unwrap_or(&[])
            {
                let local = base.join(
                    entry
                        .get("local_path")
                        .and_then(|v| v.as_str())
                        .unwrap_or(""),
                );
                let rel = local.to_string_lossy().to_string();
                if !local.exists() {
                    drift.media_missing.push(rel);
                    continue;
                }
                let Some(recorded) = entry.get("sha256").and_then(|v| v.as_str()) else {
                    continue;
                };
                if stream_sha256(&local)? != recorded {
                    drift.media_hash_mismatch.push(rel);
                } else {
                    drift.media_ok += 1;
                }
            }
        }
    }
    Ok(())
}

/// Sorted `*-examples/references/*/reference.json` under the working directory.
fn reference_records() -> Vec<PathBuf> {
    let mut out = Vec::new();
    for catalog in sorted_example_dirs() {
        let refs_dir = catalog.join("references");
        let Ok(entries) = std::fs::read_dir(&refs_dir) else {
            continue;
        };
        for entry in entries.filter_map(|e| e.ok()) {
            let candidate = entry.path().join("reference.json");
            if candidate.is_file() {
                out.push(candidate);
            }
        }
    }
    out.sort();
    out
}

/// Sorted `*-examples` directories containing a `references` directory.
fn sorted_example_dirs() -> Vec<PathBuf> {
    let mut dirs: Vec<PathBuf> = std::fs::read_dir(".")
        .into_iter()
        .flatten()
        .filter_map(|e| e.ok())
        .map(|e| {
            let p = e.path();
            p.strip_prefix(".").unwrap_or(p.as_path()).to_path_buf()
        })
        .filter(|p| {
            p.is_dir()
                && p.file_name()
                    .map(|n| n.to_string_lossy().ends_with("-examples"))
                    .unwrap_or(false)
        })
        .collect();
    dirs.sort();
    dirs
}

fn sorted_glob_examples_files(name: &str) -> Vec<PathBuf> {
    let mut files: Vec<PathBuf> = sorted_example_dirs()
        .into_iter()
        .map(|d| d.join(name))
        .filter(|p| p.is_file())
        .collect();
    files.sort();
    files
}

pub fn run(rest: &[String]) -> Result<()> {
    let mut skip_network = false;
    let mut skip_readme = false;
    let mut write_report = false;
    let mut strict = false;
    for arg in rest {
        match arg.as_str() {
            "--skip-network" => skip_network = true,
            "--skip-readme" => skip_readme = true,
            "--write-report" => write_report = true,
            "--strict" => strict = true,
            other => anyhow::bail!("unknown argument: {other}"),
        }
    }

    let mut drift = Drift::default();
    check_local_media(&mut drift)?;
    if !skip_network {
        let shared = Mutex::new(Drift::default());
        if !skip_readme {
            check_readmes(&shared)?;
        }
        check_sources(&shared);
        let net = shared.into_inner();
        drift.readme_changed = net.readme_changed;
        drift.readme_unreachable = net.readme_unreachable;
        drift.readme_unchanged = net.readme_unchanged;
        drift.sources_gone = net.sources_gone;
        drift.sources_guarded = net.sources_guarded;
        drift.sources_unresolved = net.sources_unresolved;
        drift.sources_ok = net.sources_ok;
        drift.sources_skipped = net.sources_skipped;
    }

    println!("local media verified: {}", drift.media_ok);
    println!("local media missing: {}", drift.media_missing.len());
    println!(
        "local media hash mismatch: {}",
        drift.media_hash_mismatch.len()
    );
    if !skip_network {
        println!("README snapshots unchanged: {}", drift.readme_unchanged);
        println!(
            "README snapshots changed upstream: {}",
            drift.readme_changed.len()
        );
        println!(
            "README snapshots unreachable: {}",
            drift.readme_unreachable.len()
        );
        println!("upstream URLs reachable: {}", drift.sources_ok);
        println!("upstream URLs gone: {}", drift.sources_gone.len());
        println!(
            "upstream URLs guarded (auth, rate limit, bot wall): {}",
            drift.sources_guarded.len()
        );
        println!(
            "upstream URLs unresolved (network): {}",
            drift.sources_unresolved.len()
        );
    }
    for item in drift.media_missing.iter().take(20) {
        println!("  missing media: {item}");
    }
    for item in drift.media_hash_mismatch.iter().take(20) {
        println!("  hash mismatch: {item}");
    }
    for item in drift.readme_changed.iter().take(20) {
        println!(
            "  README changed: {} ({})",
            item["repository"].as_str().unwrap_or("?"),
            item["snapshot"].as_str().unwrap_or("null"),
        );
    }
    for item in &drift.sources_gone {
        println!(
            "  upstream gone: {} -> {} {}",
            item["where"].as_str().unwrap_or("?"),
            item["result"],
            item["url"].as_str().unwrap_or("?")
        );
    }
    for item in &drift.sources_unresolved {
        println!(
            "  upstream unresolved: {} -> {} {}",
            item["where"].as_str().unwrap_or("?"),
            item["result"],
            item["url"].as_str().unwrap_or("?")
        );
    }

    if write_report {
        let report = json!({
            "schema": SCHEMA,
            "checked_at": lib::now_iso_utc(),
            "network_checked": !skip_network,
            "local_media_verified": drift.media_ok,
            "local_media_missing": drift.media_missing,
            "local_media_hash_mismatch": drift.media_hash_mismatch,
            "readme_unchanged": drift.readme_unchanged,
            "readme_changed": drift.readme_changed,
            "readme_unreachable": drift.readme_unreachable,
            "upstream_urls_reachable": drift.sources_ok,
            "upstream_urls_gone": drift.sources_gone,
            "upstream_urls_guarded": drift.sources_guarded,
            "upstream_urls_unresolved": drift.sources_unresolved,
            "upstream_urls_skipped": drift.sources_skipped,
        });
        std::fs::write(REPORT, serde_json::to_string_pretty(&report)? + "\n")?;
        println!("\nreport written to {REPORT}");
    }

    if strict && drift.any_drift() {
        // The Python original exited 1 silently; main.rs prints "error:" for
        // Err returns, so exit directly.
        std::process::exit(1);
    }
    Ok(())
}
